x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality_panel/data_quality_details/indices_details/pattern/hooks/use_ilm_explain/index.test.tsx

```diff
@@ -86,125 +88,86 @@ describe('useIlmExplain', () => {
   });
 
   describe('successful response from the ilm api', () => {
-    let ilmExplainResult: UseIlmExplain | undefined;
-
-    beforeEach(async () => {
+    function setup() {
       mockHttpFetch.mockResolvedValue(mockIlmExplain);
-
-      const { result } = renderHook(() => useIlmExplain(pattern), {
+      return renderHook(() => useIlmExplain(pattern), {
         wrapper: ContextWrapper,
       });
-      await waitFor(() => new Promise((resolve) => resolve(null)));
-      ilmExplainResult = await result.current;
-    });
+    }
 
     test('it returns the expected ilmExplain map', async () => {
-      expect(ilmExplainResult?.ilmExplain).toEqual(mockIlmExplain);
+      const { result } = setup();
+      await waitFor(() => {
+        expect(result.current.ilmExplain).toEqual(mockIlmExplain);
+      });
     });
 
     test('it returns loading: false, because the data has loaded', async () => {
-      expect(ilmExplainResult?.loading).toBe(false);
+      const { result } = setup();
+      await waitFor(() => {
+        expect(result.current.loading).toBe(false);
```

Comment:
the loading is `false` by default, I wonder if the test will fail if the loader is (incorrectly) set to true later.

Reply:
actually `loading` is `true` by default here but I get your point. Indeed in case of true -> false -> true -> false
this will not work. `waitFor` waits for an event loop tick and evaluates a first non-erroring callback passed into and then exits. So if by that time, i.e. by the time after first render the current.loading is true it will retry until it's false and then exit. But if after that false becomes true again, that will not be caught by this assertion.

Reply:
but even without waitFor if you wait a little and then evaluate you still have the same problem, if somebody comes and adds a setState a little later your test will still not catch that. It's basically a continuous catch up game because you don't know how further you should wait for with every new change.

Reply:
Ops yeah, I assumed `loading` was false by default. This is not a good example.
Let's use `error`, which is `null` by default, and there are some cases as well.
If we check `error` to be `null` inside the _waitFor_, it will evaluate `true` in the first check and will exit the, so if it (incorrectly) turns non-null later the test won't catch it.
Instead, we should _waitFor_ the request to finish, and then check the `error` is still `null`.
Does it make sense?

Reply:
done for use_ilm_explain and use_stats tests. Thanks